### ISSUE TITLE
Unnecessary delays in SQS?

### ISSUE THREAD

**Issue by elliotsayes** [ar-io/arweave-gateway#5]
Currently small arweave transactions submitted to the arweave.net gateway take upwards of two minutes to become accessible for download from the same gateway. I am investigating if it is possible to reduce this.

Look at the gateway source code, it seems that delays in SQS are responsible for much of the delay, e.g.
https://github.com/ar-io/arweave-gateway/blob/385ac37feca16e16af6d8095acaa3a90babc59bf/terraform/src/sqs.tf#L52
It seems that these delays are simply defaults taken from: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/sqs_queue#fifo-queue

I would appreciate if someone more familiar with the architecture could check if it is safe to reduce/remove these delays. If so I'll be happy to make a PR

Further discussion here: https://discord.com/channels/908759493943394334/921474264387964938/1039707967693525042

**Comment by hlolli** [comment]
this delay is intended, this specific queue is export-chunks. The 90 seconds delay is to make sure the nodes backing the gateway have received the tx headers which will be matched up against the chunk hashes. This has 0 effect on the user experience since we optimistically show the data as soon as they arrive, the exporting of the chunks has to do with moving the chunks from the gateway into the arweave network proper. It's important this succeeds and less important for the user who's uploading data, to wait a little bit for the chunks to be retrieveable from the nodes. It will take few days for chunks to be spread troughout the network anyway. So the 90 second delay prevents errors such has "data_root_not_found" which is a result of sending them too early.

**Comment by elliotsayes** [comment]
I see, so that particular queue is responsible for sending data to the arweave nodes proper. In that case it seems I’ve misidentified the purpose of that queue, and a delay is indeed reasonable as it is only a small component of the critical path.

However, as mentioned in the issues, I’m interested in how long it takes for the transaction data to be publicly downloadable from arweave.net via HTTPS. The fact remains that this process takes upwards of two minutes. 

As you sat, it is not caused by this particular queue, so then is it caused by another of the queues defined in that same file? Or if it is not caused by **any** of the queues, then what else could be causing the delay?

**Comment by hlolli** [comment]
it depends a lot if it's 1. a bundle, 2. a manifest. It's most likely a mix of TTL on 404 response and all the chunks to be written into s3 bucket and then assembled on demand.

**Comment by elliotsayes** [comment]
In that case, the low hanging fruit seems like it would be something like: 
`res.set('Cache-Control', 'no-store')`
on 404s for the txId route. 

I’ll take a look into how saving the chunks to s3 works.

**Comment by hlolli** [comment]
we have ddos attacks regularly and we just can't do it

**Comment by hlolli** [comment]
I would tell you what TTL we have, but that could also be exploited. But this isn't about browser cache, it's more to do with cloudfront.

**Comment by elliotsayes** [comment]
Damn, that is annoying 😕. And it seems like Cloudfront ignores the cache control header for 404s anyway. 

https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/HTTPStatusCodes.html#HTTPStatusCodes-cached-errors

**Comment by hlolli** [comment]
yes true, but this is not where it's set, rather in custom error response. But that value is way less than 2 minutes. The elevator pitch for the time for data to be available is: if you upload a bundle, it needs to be unbundled and that takes time, if you upload a small layer-1 data transaction, it will be available almost immediately.

**Comment by elliotsayes** [comment]
Got it, that makes sense. Looks like this won't be as simple as I'm making it out to be.

Just for reference, the transactions I have been testing with are not bundles. However the data part of the Transaction is uploaded separately to the `chunk` endpoint (even though the data is small enough that it could still fit in the transaction). I will do some testing and see if the delay could be related to that.

**Comment by elliotsayes** [comment]
I've confirmed that the issue seems to be unique to transactions uploaded via chunks. See test code below:

```js
const filePath = 'TEST_DATA.txt'
const fileData = fs.readFileSync(filePath, 'utf-8')

const submitStandaloneTx = async () => {
  const tx = await arweave.createTransaction({
    data: fileData,
  }, jwk)
  await arweave.transactions.sign(tx, jwk)
  const res = await arweave.transactions.post(tx);
  return {type: 'standalone', txId: tx.id}
}

const submitChunkedTx = async () => {
  const tx = await pipeline(fs.createReadStream(filePath), arweaveStreamTx.createTransactionAsync({}, arweave, jwk));
  await arweave.transactions.sign(tx, jwk);
  await pipeline(fs.createReadStream(filePath), arweaveStreamTx.uploadTransactionAsync(tx, arweave));
  return {type: 'chunked', txId: tx.id}
}

// this transaction appears very quickly at arweave.net/{txId}
submitStandaloneTx().then(res => console.log(res))

// this transaction takes a couple minutes to appear at arweave.net/{txId}
submitChunkedTx().then(res => console.log(res))
```

**Comment by elliotsayes** [comment]
In this case could it be something to do with the delay in the `import-chunks` queue, rather than `export-chunks`? The `import-chunks` handler seems to simply write the chunk confirmation into the database. Is it still possible to have a race condition here? It's possible that there is a race condition with the `import-txs` queue, but that looks like it has an identical 90 second delay anyway 🤔

Looking at the chunk table accessed by `import-chunks` there don't appear to be any foreign keys constraints that might make race conditions unsafe

https://github.com/ar-io/arweave-gateway/blob/main/sql/gateway-schema.sql#L69

**Comment by hlolli** [comment]
good catch, there's indeed a `delay_seconds = 90` in import-chunks in this repo as well as in our private repo. This can probably be brought down. That said, we have a feature in place that's missing in this public repo I believe, which is that we optimistically look up all the chunks belonging to a data_root within a folder and check if it's all in the s3 bucket, and then assemble them on demand. But if it's true what you say, that perhaps this fallback mechanism isn't properly kicking in, then it would surely help to reduce the delay time on import-chunks. Another mechanism also missing in this repository is a lamba+sqs queue which assembles chunks into contiguous data files even before the user requests them, this we did in order to reduce response times, which are generally longer when we are assembling the chunks on demand.
But I think the takeaway from your observation is that we should reduce the delay in import-chunks and see what happens, I can notify you when we have that on prod and you could do another test?

**Comment by elliotsayes** [comment]
That would be great @hlolli, appreciate that you are open to doing this

**Comment by elliotsayes** [comment]
Is this likely to be deployed in the next couple of days ? Planning a launch and would be great to have ready by then :)

**Comment by hlolli** [comment]
> Is this likely to be deployed in the next couple of days ? Planning a launch and would be great to have ready by then :)

it's been since right now set from 90 to 0, maybe it helps, hard to tell

**Comment by elliotsayes** [comment]
Don't think it's 90s faster. Some approximate delays for chunked txs (including 404 caching):
`POyMbnW0unPVUMCbDwvQlsKti6vNoJlE8YZgk_kXNCI` ~2:15
`jJJ2TaCbQSQ16yVZmo528fmtLFQcjyGOFpDwgeWFGog` ~2:10
